File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/NewsFragment.java
package com.pennapps.labs.pennmobile;

import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.customtabs.CustomTabsClient;
import android.support.customtabs.CustomTabsIntent;
import android.support.customtabs.CustomTabsService;
import android.support.customtabs.CustomTabsServiceConnection;
import android.support.customtabs.CustomTabsSession;
import android.support.v4.app.ListFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import com.crashlytics.android.Crashlytics;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.ContentViewEvent;

import java.util.ArrayList;
import java.util.List;

import butterknife.ButterKnife;
import io.fabric.sdk.android.Fabric;

public class NewsFragment extends ListFragment {

    private ListView mListView;
    private CustomTabsClient mCustomTabsClient;
    private CustomTabsIntent customTabsIntent;
    private Intent share;
    private CustomTabsSession session;
    private CustomTabsIntent.Builder builder;
    private boolean isCustomTabsSupported;

    class CustomListAdapter extends ArrayAdapter<String> {

        private final Context context;
        private NewsSite[] news;

        public CustomListAdapter(Context context, String[] newsNames, NewsSite[] news) {
            super(context, R.layout.fragment_news, newsNames);
            this.news = news;
            this.context=context;
        }

        public View getView(int position,View view,ViewGroup parent) {
            LayoutInflater inflater = LayoutInflater.from(context);
            View rowView=inflater.inflate(R.layout.news_list_item, null,true);

            TextView newsName = (TextView) rowView.findViewById(R.id.news_name);
            ImageView newsLogo = (ImageView) rowView.findViewById(R.id.news_logo);
            TextView newsDetails = (TextView) rowView.findViewById(R.id.news_details);

            newsName.setText(news[position].getName());
            newsLogo.setImageResource(news[position].getImage());
            newsDetails.setText(news[position].getDescription());
            return rowView;

        }
    }

    class NewsSite {
        private String name, url, description;
        private int image;


        public NewsSite(String name, String url, String description, int image) {
            this.name = name;
            this.url = url;
            this.image = image;
            this.description = description;
        }

        public String getName() {
            return name;
        }

        public int getImage() {
            return image;
        }

        public String getUrl() {
            return url;
        }

        public String getDescription() { return description; }

        @Override
        public String toString() {
            return name;
        }
    }

    class NewsCustomTabsServiceConnection extends CustomTabsServiceConnection {
        public static final String CUSTOM_TAB_PACKAGE_NAME = "com.android.chrome";
        @Override
        public void onCustomTabsServiceConnected(ComponentName name, CustomTabsClient client) {
            mCustomTabsClient = client;
            mCustomTabsClient.warmup(0);
            session = mCustomTabsClient.newSession(null);
            final ArrayList<String> URLs = new ArrayList<>();
            ArrayList<String> titles = new ArrayList<>();
            URLs.add("http://www.thedp.com/");
            URLs.add("http://www.34st.com/");
            URLs.add("http://www.thedp.com/blog/under-the-button/");
            titles.add("The Daily Pennsylvanian");
            titles.add("34th Street");
            titles.add("Under the Button");
            ArrayList<Bundle> urlList = new ArrayList<>();
            for (int i = 0; i < URLs.size(); i++) {
                Bundle bundle = new Bundle();
                bundle.putParcelable(CustomTabsService.KEY_URL, new Parcelable() {
                    @Override
                    public int describeContents() {
                        return 0;
                    }

                    @Override
                    public void writeToParcel(Parcel parcel, int i) {
                        parcel.writeString(URLs.get(i));
                    }
                });
                urlList.add(bundle);
            }
            session.mayLaunchUrl(Uri.parse(URLs.get(0)), null, urlList);
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            mCustomTabsClient = null;
            session = null;
            customTabsIntent = null;
        }
    }

    private static boolean isChromeCustomTabsSupported(final Context context) {
        String SERVICE_ACTION = "android.support.customtabs.action.CustomTabsService";
        Intent serviceIntent = new Intent(SERVICE_ACTION);
        serviceIntent.setPackage("com.android.chrome");
        List<ResolveInfo> resolveInfos = context.getPackageManager().queryIntentServices(serviceIntent, 0);
        return !(resolveInfos == null || resolveInfos.isEmpty());
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        CustomTabsServiceConnection connection = new NewsCustomTabsServiceConnection();
        isCustomTabsSupported = isChromeCustomTabsSupported(getContext());
        setHasOptionsMenu(true);
        mListView = getListView();
        builder = new CustomTabsIntent.Builder();
        share = new Intent(Intent.ACTION_SEND);
        share.setType("text/plain");
        builder.setToolbarColor(0x3E50B4);
        builder.setStartAnimations(getContext(),
                android.support.design.R.anim.abc_popup_enter,
                android.support.design.R.anim.abc_popup_exit);

        CustomTabsClient.bindCustomTabsService(getContext(),
                NewsCustomTabsServiceConnection.CUSTOM_TAB_PACKAGE_NAME, connection);
        addNews();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((MainActivity) getActivity()).closeKeyboard();
        setHasOptionsMenu(true);

        Fabric.with(getContext(), new Crashlytics());
        Answers.getInstance().logContentView(new ContentViewEvent()
                .putContentName("News")
                .putContentType("App Feature")
                .putContentId("5"));
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_news, container, false);
        ButterKnife.bind(this, v);
        return v;
    }

    private void addNews() {
        String dpDescription = "The Daily Pennsylvanian is the independent student newspaper of the University of Pennsylvania.";
        String thirtyFourDescription = "34th Street Magazine is the DP's arts and entertainment weekly magazine.";
        String utbDescription = "Under The Button is Penn's 24/7 news and entertainment blog, known for its signature humor, gossip and snarky features.";
        NewsSite dp = new NewsSite("The Daily Pennsylvanian", "http://www.thedp.com/",
                dpDescription, R.drawable.thedp);
        NewsSite thirtyFour = new NewsSite("34th Street", "http://www.34st.com/",
                thirtyFourDescription, R.drawable.thirtyfour);
        NewsSite utb = new NewsSite("Under the Button",
                "https://www.underthebutton.com/", utbDescription, R.drawable.utb);
        NewsSite[] allSites = {dp, thirtyFour, utb};
        String[] newsUrls = new String[allSites.length];
        for (int i = 0; i < newsUrls.length; i++) {
            newsUrls[i] = allSites[i].getUrl();
        }
        CustomListAdapter adapter = new CustomListAdapter(getContext(), newsUrls, allSites);
        mListView.setAdapter(adapter);
    }

    @Override
    public void onListItemClick(ListView l, View v, int position, long id) {
        String url = (String) l.getItemAtPosition(position);
        if (url != null) {
            if (isCustomTabsSupported) {
                share.putExtra(Intent.EXTRA_TEXT, url);
                builder.addMenuItem("Share", PendingIntent.getActivity(getContext(), 0,
                        share, PendingIntent.FLAG_CANCEL_CURRENT));
                customTabsIntent = builder.build();
                customTabsIntent.launchUrl(getActivity(), Uri.parse(url));
            } else {
                    Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
                    startActivity(browserIntent);
            }
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        getActivity().setTitle(R.string.news);
        ((MainActivity) getActivity()).setNav(R.id.nav_news);
    }

    @Override
    public void onDestroyView() {
        ((MainActivity) getActivity()).removeTabs();
        super.onDestroyView();

    }
}

File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/NewsFragment.kt
package com.pennapps.labs.pennmobile

import android.app.PendingIntent
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.res.Configuration
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import androidx.browser.customtabs.*
import androidx.fragment.app.ListFragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import androidx.preference.PreferenceManager
import com.google.firebase.analytics.FirebaseAnalytics
import kotlinx.android.synthetic.main.fragment_news.*
import kotlinx.android.synthetic.main.fragment_news.initials
import kotlinx.android.synthetic.main.fragment_news.profile_background
import java.util.ArrayList


class NewsFragment : ListFragment() {

    companion object {
        @JvmStatic val CUSTOM_TAB_PACKAGE_NAME = "com.android.chrome"
    }

    private var mListView: ListView? = null
    private var mCustomTabsClient: CustomTabsClient? = null
    private var customTabsIntent: CustomTabsIntent? = null
    private var share: Intent? = null
    private var session: CustomTabsSession? = null
    private var builder: CustomTabsIntent.Builder? = null
    private var isCustomTabsSupported: Boolean = false
    private lateinit var sharedPreferences: SharedPreferences

    internal inner class CustomListAdapter(@get:JvmName("getContext_") private val context: Context,
                                           newsNames: Array<String?>,
                                           private val news: Array<NewsSite>) : ArrayAdapter<String>(context, R.layout.fragment_news, newsNames) {

        override fun getView(position: Int, view: View?, parent: ViewGroup): View {
            val inflater = LayoutInflater.from(context)
            val rowView = inflater.inflate(R.layout.news_list_item, null, true)

            val newsName = rowView.findViewById(R.id.news_name) as TextView
            val newsLogo = rowView.findViewById(R.id.news_card_logo) as ImageView
            val newsDetails = rowView.findViewById(R.id.news_details) as TextView

            newsName.text = news[position].name
            newsLogo.setImageResource(news[position].image)
            newsDetails.text = news[position].description
            return rowView
        }
    }

    internal inner class NewsSite(val name: String, val url: String, val description: String, val image: Int) {

        override fun toString() = name
    }

    internal inner class NewsCustomTabsServiceConnection : CustomTabsServiceConnection() {

        override fun onCustomTabsServiceConnected(name: ComponentName, client: CustomTabsClient) {
            mCustomTabsClient = client
            mCustomTabsClient?.warmup(0)
            session = mCustomTabsClient?.newSession(null)
            val URLs = ArrayList<String>()
            val titles = ArrayList<String>()
            URLs.add("http://www.thedp.com/")
            URLs.add("http://www.34st.com/")
            URLs.add("https://www.underthebutton.com/")
            titles.add("The Daily Pennsylvanian")
            titles.add("34th Street")
            titles.add("Under the Button")
            val urlList = ArrayList<Bundle>()
            for (i in URLs.indices) {
                val bundle = Bundle()
                bundle.putParcelable(CustomTabsService.KEY_URL, object : Parcelable {
                    override fun describeContents() = 0

                    override fun writeToParcel(parcel: Parcel, i: Int) { parcel.writeString(URLs[i]) }
                })
                urlList.add(bundle)
            }
            session?.mayLaunchUrl(Uri.parse(URLs[0]), null, urlList)
        }

        override fun onServiceDisconnected(name: ComponentName) {
            mCustomTabsClient = null
            session = null
            customTabsIntent = null
        }
    }

    private fun isChromeCustomTabsSupported(context: Context): Boolean {
        val SERVICE_ACTION = "android.support.customtabs.action.CustomTabsService"
        val serviceIntent = Intent(SERVICE_ACTION)
        serviceIntent.setPackage("com.android.chrome")
        val resolveInfos = context.packageManager.queryIntentServices(serviceIntent, 0)
        return resolveInfos.isNotEmpty()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val connection = NewsCustomTabsServiceConnection()
        context?.let { context ->
            isCustomTabsSupported = isChromeCustomTabsSupported(context)
        }
        setHasOptionsMenu(true)
        mListView = listView
        builder = CustomTabsIntent.Builder()
        share = Intent(Intent.ACTION_SEND)
        share?.type = "text/plain"
        builder?.setToolbarColor(0x3E50B4)
        context?.let { context ->
            builder?.setStartAnimations(context,
                    R.anim.abc_popup_enter,
                    R.anim.abc_popup_exit)
            CustomTabsClient.bindCustomTabsService(context,
                    CUSTOM_TAB_PACKAGE_NAME, connection)
        }

        addNews()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val mActivity = activity as MainActivity
        mActivity.closeKeyboard()
        setHasOptionsMenu(true)

        val bundle = Bundle()
        bundle.putString(FirebaseAnalytics.Param.ITEM_ID, "5")
        bundle.putString(FirebaseAnalytics.Param.ITEM_NAME, "News")
        bundle.putString(FirebaseAnalytics.Param.ITEM_CATEGORY, "App Feature")
        FirebaseAnalytics.getInstance(mActivity).logEvent(FirebaseAnalytics.Event.VIEW_ITEM, bundle)
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(mActivity)

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_news, container, false)
        return view
    }

    private fun addNews() {

        //displays banner if not connected
        if (!isOnline(context)) {
            internetConnectionNews?.setBackgroundColor(resources.getColor(R.color.darkRedBackground))
            internetConnection_message_news?.text = resources.getString(R.string.internet_error)
            internetConnectionNews?.visibility = View.VISIBLE
        } else {
            internetConnectionNews?.visibility = View.GONE
        }

        val dpDescription = "The Daily Pennsylvanian is the independent student newspaper of the University of Pennsylvania."
        val thirtyFourDescription = "34th Street Magazine is the DP's arts and entertainment weekly magazine."
        val utbDescription = "Under The Button is Penn's 24/7 news and entertainment blog, known for its signature humor, gossip and snarky features."
        val dp = NewsSite("The Daily Pennsylvanian", "http://www.thedp.com/",
                dpDescription, R.drawable.thedp)
        var thirtyFour = NewsSite("34th Street", "http://www.34st.com/",
                thirtyFourDescription, R.drawable.thirtyfour)
        var utb = NewsSite("Under the Button",
                "https://www.underthebutton.com/", utbDescription, R.drawable.utb)

        if (Build.VERSION.SDK_INT > 28 && (resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_YES) {
            thirtyFour = NewsSite("34th Street", "http://www.34st.com/",
                    thirtyFourDescription, R.drawable.thirtyfour_darkmode)
            utb = NewsSite("Under the Button",
                    "https://www.underthebutton.com/", utbDescription, R.drawable.utb_darkmode)
        }

        val allSites = arrayOf(dp, thirtyFour, utb)
        val newsUrls = arrayOfNulls<String>(allSites.size)
        for (i in newsUrls.indices) {
            newsUrls[i] = allSites[i].url
        }
        context?.let {context ->
            val adapter = CustomListAdapter(context, newsUrls, allSites)
            mListView?.setAdapter(adapter)
        }


    }


    override fun onListItemClick(l: ListView, v: View, position: Int, id: Long) {
        l.let {l ->
            val url = l.getItemAtPosition(position) as String
                if (isCustomTabsSupported) {
                    share?.putExtra(Intent.EXTRA_TEXT, url)
                    builder?.addMenuItem("Share", PendingIntent.getActivity(context, 0,
                            share, PendingIntent.FLAG_CANCEL_CURRENT))
                    customTabsIntent = builder?.build()
                    activity?.let { activity ->
                        customTabsIntent?.launchUrl(activity, Uri.parse(url))
                    }
                } else {
                    val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
                    startActivity(browserIntent)
                }
            }
      }

    override fun onResume() {
        super.onResume()
        val mActivity : MainActivity? = activity as MainActivity
        mActivity?.removeTabs()
        mActivity?.setTitle(R.string.news)
        val initials = sharedPreferences.getString(getString(R.string.initials), null)
        if (initials != null && initials.isNotEmpty()) {
            this.initials.text = initials
        } else {
            this.profile_background.setImageDrawable(
                    resources.getDrawable
                    (R.drawable.ic_guest_avatar, context?.theme))
        }
        mActivity?.setSelectedTab(MainActivity.MORE)
    }

    override fun onDestroyView() {
        val mActivity : MainActivity? = activity as MainActivity
        mActivity?.removeTabs()
        super.onDestroyView()
    }
}